src/enzo/ActiveParticle_SmartStar.h

```diff
@@ -0,0 +1,450 @@
+  static int EvaluateFeedback(
+      grid *thisgrid_orig, ActiveParticleFormationData &data);
+  template <class active_particle_class>
+    static int BeforeEvolveLevel(HierarchyEntry *Grids[], TopGridData *MetaData,
+				 int NumberOfGrids, LevelHierarchyEntry *LevelArray[], 
+				 int ThisLevel, bool CallEvolvePhotons,
+				 int TotalStarParticleCountPrevious[],
+				 int SmartStarID);
+  template <class active_particle_class>
+    static int AfterEvolveLevel(HierarchyEntry *Grids[], TopGridData *MetaData,
+				int NumberOfGrids, LevelHierarchyEntry *LevelArray[], 
+				int ThisLevel, int TotalStarParticleCountPrevious[],
+				int SmartStarID);
+  template <class active_particle_class>
+    static int DepositMass(HierarchyEntry *Grids[], TopGridData *MetaData,
+				int NumberOfGrids, LevelHierarchyEntry *LevelArray[], 
+				int ThisLevel, int GalaxyParticleID) {return SUCCESS; };
+  static int SetFlaggingField(
+      LevelHierarchyEntry *LevelArray[], 
+      int level, 
+      int TopGridDims[], 
+      int ActiveParticleID);
+
+  static int SmartStarParticleFeedback(
+             int nParticles, 
+             ActiveParticleList<ActiveParticleType>& ParticleList,
+		     FLOAT dx, LevelHierarchyEntry *LevelArray[], int ThisLevel);
+  
+
+  static int ResetAcceleration(float *ActiveParticleAcceleration);
+  static int CreateParticle(grid *thisgrid_orig, ActiveParticleFormationData &supp_data,
+			    int particle_index);
+  static int InitializeParticleType();
+  void   SmartMerge(ActiveParticleType_SmartStar *a);
+  int CalculateAccretedAngularMomentum();
+  int SmartStarAddFeedbackSphere();
+  ENABLED_PARTICLE_ID_ACCESSOR
+  bool IsARadiationSource(FLOAT Time);
+  
+  // sink helper routines
+
+  template <class active_particle_class>
+  static void MergeSmartStars(
+      int *nParticles, ActiveParticleList<ActiveParticleType>& ParticleList, 
+      int *ngroups, 
+      LevelHierarchyEntry *LevelArray[], int ThisLevel, 
+      ActiveParticleList<active_particle_class>& MergedParticles);
+
+  static int Accrete(int nParticles, 
+      ActiveParticleList<ActiveParticleType>& ParticleList,
+      FLOAT AccretionRadius, FLOAT dx, 
+      LevelHierarchyEntry *LevelArray[], int ThisLevel);
+
+  static int UpdateAccretionRateStats(int nParticles,
+				      ActiveParticleList<ActiveParticleType>& ParticleList,
+				      FLOAT dx, 
+				      LevelHierarchyEntry *LevelArray[], int ThisLevel);
+
+ 
+  static float EjectedMassThreshold;
+  FLOAT AccretionRadius;   // in units of CellWidth on the maximum refinement level
+  static int RadiationParticle;
+  static double LuminosityPerSolarMass;
+  static int RadiationSEDNumberOfBins;
+  static float* RadiationEnergyBins;
+  static float* RadiationSED;
+  static float RadiationLifetime;
+  //float acc[3];
+  int ParticleClass;
+
+  float AccretionRate[NTIMES];  //In would be nice to make this a vector
+  //std::vector<float> AccretionRateTime;
+  float AccretionRateTime[NTIMES];
```

Comment:
If it makes things better, these should be made into vectors before merging.

Reply:
I tried that before but it wasn't compatible with the way the AP infrastructure is set up. Having the AccretionRate and AccretionRateTime as arrays works very nicely anyway.